README.md
[Ember CLI Deploy Rack]
=======================

[![Travis CI Status][Travis CI Status]][Travis CI]
[![Code Climate Status][Code Climate Status]][Code Climate]
[![Test Coverage Status][Test Coverage Status]][Test Coverage]
[![Gemnasium Status][Gemnasium Status]][Gemnasium]
[![Gem Version][Gem Version]][RubyGems]

**A [Rack] package to launch your [Ember.js] application into the Cloud.**

Summary
-------

Ember CLI Deploy Rack can be used to serve an [Ember CLI] application deployed with the help of [Ember CLI Deploy] and
[ember-deploy-redis].

Basically, it is a [Sinatra] application, which connects to Redis and returns the uploaded index.html.

You can deploy your application to production, test it out by passing an `revision` parameter with the revision you want
to test and activate when you feel confident that everything is working as expected.

For further information see [Ember CLI Deploy].

Prerequisites
-------------

* [Ruby]
* [Redis]

Installation
------------

1. Create a Gemfile

    ```ruby
    #!/usr/bin/env bundle

    source 'https://rubygems.org'

    gem 'ember-cli-deploy-rack', require: false
    ```

2. Run `bundle` to install all dependencies with [Bundler]

Usage
-----

1. Create a Rack configuration

    ```ruby
    #!/usr/bin/env rackup

    # This file is used by Rack-based servers to start the application.

    require 'bundler/setup'

    require 'ember/cli/deploy/rack'

    run Ember::CLI::Deploy::Rack::Engine
    ```

2. Run the engine with `rackup` or your favorite Ruby Webserver

Configuration
-------------

Ember CLI Deploy Rack can be configured in various ways:

1. Via settings

    ```ruby
    engine = Ember::CLI::Deploy::Rack::Engine.new

    # === Settings ====

    # engine.settings.set :root, Dir.pwd
    # engine.settings.set :key_prefix, 'ember-cli-deploy-rack:index'
    # engine.settings.set :active_content_suffix, 'current-content'
    # engine.settings.set :revision, { regexp: '^[0-9a-f]{32}$' }
    # engine.settings.set :redis, { host: '127.0.0.1', ... }
    # engine.settings.set :redis_client, proc { Redis.new settings.redis }
    # engine.settings.set :debug, false
    # ...

    # === Logging ====

    # log_dir = File.expand_path 'log'
    #
    # if File.directory? log_dir
    #   log_file = File.expand_path "#{engine.settings.environment}.log", log_dir
    #
    #   logger = Logger.new log_file, File::WRONLY | File::APPEND | File::CREAT
    #   logger.level = Logger::ERROR
    #
    #   engine.settings.set :logger, logger
    # end

    run engine
    ```

2. Via YAML configuration file

    Just create a configuration file `config/settings.yml` and adjust properly, for an example see
    [resources/config/settings.yml].

### Options

#### Key Prefix

The prefix to be used for the Redis key under which file will be uploaded to Redis. The Redis key will be a combination
of the `key_prefix` and the `revision_key`. Typically this is the `project.name` property from the deployment context
and `:index` at the end. By default this option will be `ember-cli-deploy-rack:index`.

For more details, see https://github.com/ember-cli-deploy/ember-cli-deploy-redis#keyprefix.

#### Active Content Suffix

The suffix to be used for the Redis key under which the activated revision content will be stored in Redis. By default
this option will be `current-content`. This makes the default activated revision in Redis looks like: The `project.name`
property from the deployment and `:index:current-content` at the end.

For more details, see https://github.com/ember-cli-deploy/ember-cli-deploy-redis#activecontentsuffix.

#### Revision

#### Regexp

The regular expression to check the revision. By default this option will be `^[0-9a-f]{32}$`.

#### Redis

The options for the Redis client. For available options see [Redis::Client::DEFAULTS].

#### Debug

If enabled, debug information can be viewed under `/debug`. By default this option will be `false`.

Development
-----------

### Run specs with [RSpec]

Run `rspec`.

or via [Guard]:

```
$ guard -g spec
```

### See Test Coverage

Run `COVERAGE=true rspec`.

### Run [RuboCop]

Run `rubocop`.

To run all specs and RuboCop altogether, run `rake`.

### Rack

```
$ rackup
```

#### Thin

```
$ thin -R config.ru -d start
$ thin stop
```

Bug Reports
-----------

GitHub Issues are used for managing bug reports and feature requests. If you run into issues, please search the issues
and submit new problems [here].

Versioning
----------

This library aims to adhere to [Semantic Versioning 2.0.0]. Violations of this scheme should be reported as bugs.
Specifically, if a minor or patch version is released that breaks backward compatibility, that version should be
immediately yanked and / or a new version should be immediately released that restores compatibility.

License
-------

Ember CLI Deploy Rack is released under the [MIT License (MIT)], see [LICENSE].

[Bundler]: http://bundler.io "The best way to manage a Ruby application's gems"
[Code Climate]: https://codeclimate.com/github/bitaculous/ember-cli-deploy-rack "Ember CLI Deploy Rack at Code Climate"
[Code Climate Status]: https://img.shields.io/codeclimate/github/bitaculous/ember-cli-deploy-rack.svg?style=flat "Code Climate Status"
[Ember CLI]: http://www.ember-cli.com "The command line interface for ambitious web applications."
[Ember CLI Deploy Rack]: https://bitaculous.github.io/ember-cli-deploy-rack/ "A Rack package to launch your Ember.js application into the Cloud."
[Ember CLI Deploy]: http://ember-cli.github.io/ember-cli-deploy "Simple, flexible deployment for your Ember app"
[ember-deploy-redis]: https://github.com/ember-cli-deploy/ember-cli-deploy-redis "An ember-cli-deploy plugin to upload index.html to a Redis store."
[Ember.js]: http://emberjs.com "A framework for creating ambitious web applications."
[Foreman]: http://ddollar.github.io/foreman "Manage Procfile-based applications"
[Gem Version]: https://img.shields.io/gem/v/ember-cli-deploy-rack.svg?style=flat "Gem Version"
[Gemnasium]: https://gemnasium.com/bitaculous/ember-cli-deploy-rack "Ember CLI Deploy Rack at Gemnasium"
[Gemnasium Status]: https://img.shields.io/gemnasium/bitaculous/ember-cli-deploy-rack.svg?style=flat "Gemnasium Status"
[Guard]: http://guardgem.org "A command line tool to easily handle events on file system modifications."
[here]: https://github.com/bitaculous/ember-cli-deploy-rack/issues "Github Issues"
[LICENSE]: https://raw.githubusercontent.com/bitaculous/ember-cli-deploy-rack/master/LICENSE "License"
[MIT License (MIT)]: http://opensource.org/licenses/MIT "The MIT License (MIT)"
[Rack]: http://rack.github.io "A Ruby Webserver Interface"
[Redis]: http://redis.io "An open source, BSD licensed, advanced key-value cache and store."
[Redis::Client::DEFAULTS]: https://github.com/redis/redis-rb/blob/master/lib/redis/client.rb#L8 "The redis-rb defaults"
[resources/config/settings.yml]: https://github.com/bitaculous/ember-cli-deploy-rack/blob/master/resources/config/settings.yml "Sample YAML configuration"
[RSpec]: http://rspec.info "Behaviour Driven Development for Ruby"
[RuboCop]: https://github.com/bbatsov/rubocop "A Ruby static code analyzer, based on the community Ruby style guide."
[Ruby]: https://www.ruby-lang.org "A dynamic, open source programming language with a focus on simplicity and productivity."
[RubyGems]: https://rubygems.org/gems/ember-cli-deploy-rack "Ember CLI Deploy Rack at RubyGems"
[Semantic Versioning 2.0.0]: http://semver.org "Semantic Versioning 2.0.0"
[Sinatra]: http://www.sinatrarb.com "A DSL for quickly creating web applications in Ruby with minimal effort."
[Test Coverage]: https://codeclimate.com/github/bitaculous/ember-cli-deploy-rack "Test Coverage (Code Climate)"
[Test Coverage Status]: https://img.shields.io/codeclimate/coverage/github/bitaculous/ember-cli-deploy-rack.svg?style=flat "Test Coverage Status"
[Travis CI]: https://travis-ci.org/bitaculous/ember-cli-deploy-rack "Ember CLI Deploy Rack at Travis CI"
[Travis CI Status]: https://img.shields.io/travis/bitaculous/ember-cli-deploy-rack.svg?style=flat "Travis CI Status"

CHANGELOG.md
2.5.0
=====

* [✓] Catch all routes, fixes #1
* [➠] Updated Gems

2.4.4
=====

* [➠] Updated Gems

2.4.3
=====

* [➠] Updated Gems

2.4.2
=====

* [➠] Added metadata to Gemspec

2.4.1
=====

* [➠] Updated Gems

2.4.0
=====

* [➠] Updated RuboCop configuration
* [➠] Updated Sinatra

2.3.2
=====

* [➠] Freeze mutable objects assigned to constants

2.3.1
=====

* [➠] Moved Rake tasks out of library
* [➠] Updated Gems

2.3.0
=====

* [✖] Removed logging out of engine, should be configured via `config.ru`
* [✓] Use `Sinatra::Base` instead of `Sinatra::Application`
* [➠] Improved `debug` view

2.2.0 (Yanked)
==============

* [➠] First release on RubyGems
* [➠] Debug information can be enabled / disabled
* [➠] Formatted debug information

2.1.0
=====

* [➠] Updated README
* [➠] Refactored engine code
* [✚] Added `debug` route
* [✖] Removed `config` gem, `sinatra/config_file` is perfectly adequate

2.0.1
=====

* [➠] Response with 400 when revision is not valid

2.0.0
=====

* [➠] Renamed `index_id` to  `key_prefix`
* [➠] Added `active_content_suffix` and `revision` options
* [➠] Compatibility with the latest version of `ember-cli-deploy-redis`

1.0.1
=====

* [✓] Fixed version require

1.0.0
=====

* [✚] Added specs
* [➠] Configured Travis CI, Code Climate, Test Coverage, RuboCop and Gemnasium
* [❶] First commit (born from Bitaculous Booster)

spec/spec_helper.rb
# Conventionally, all specs live under a `spec` directory, which RSpec adds to the `$LOAD_PATH`. The `.rspec` file
# contains `--require ember_cli_deploy_rack_helper` which will cause this file to always be loaded, without a need to
# explicitly require it in any files.
#
# Given that it is always loaded, you are encouraged to keep this file as light-weight as possible. Requiring
# heavyweight dependencies from this file will add to the boot time of your test suite on EVERY test run, even for an
# individual file that may not need all of that loaded. Instead, consider making a separate helper file that requires
# the additional dependencies and performs the additional setup, and require it from the spec files that actually need
# it.
#
# The `.rspec` file also contains a few flags that are not defaults but that users commonly want.
#
# See <http://rubydoc.info/gems/rspec-core/RSpec/Core/Configuration> for more information.

RSpec.configure do |config|
  # RSpec Expectations configuration goes here. You can use an alternate assertion / expectation library such as `wrong`
  # or the `stdlib/minitest` assertions if you prefer.
  config.expect_with :rspec do |expectations|
    # This option will default to `true` in RSpec 4. It makes the `description` and `failure_message` of custom matchers
    # include text for helper methods defined using `chain`.
    expectations.include_chain_clauses_in_custom_matcher_descriptions = true
  end

  # RSpec Mocks configuration goes here. You can use an alternate test double library (such as `bogus` or `mocha`) by
  # changing the `mock_with` option here.
  config.mock_with :rspec do |mocks|
    # Prevents you from mocking or stubbing a method that does not exist on a real object. This is generally recommended,
    # and will default to `true` in RSpec 4.
    mocks.verify_partial_doubles = true
  end

  # The settings below are suggested to provide a good initial experience with RSpec, but feel free to customize to your
  # heart's content.

  # These two settings work together to allow you to limit a spec run to individual examples or groups you care about by
  # tagging them with `:focus` metadata. When nothing is tagged with `:focus`, all examples get run.
  config.filter_run :focus
  config.run_all_when_everything_filtered = true

  # Allows RSpec to persist some state between runs in order to support the `--only-failures` and `--next-failure` CLI
  # options. We recommend you configure your source control system to ignore this file.
  # config.example_status_persistence_file_path = 'spec/examples.txt'

  # Limits the available syntax to the non-monkey patched syntax that is recommended. For more details, see:
  #
  # - <http://myronmars.to/n/dev-blog/2012/06/rspecs-new-expectation-syntax>
  # - <http://www.teaisaweso.me/blog/2013/05/27/rspecs-new-message-expectation-syntax>
  # - <http://myronmars.to/n/dev-blog/2014/05/notable-changes-in-rspec-3#new__config_option_to_disable_rspeccore_monkey_patching>
  config.disable_monkey_patching!

  # Many RSpec users commonly either run the entire suite or an individual file, and it's useful to allow more verbose
  # output when running an individual spec file.
  if config.files_to_run.one?
    # Use the documentation formatter for detailed output, unless a formatter has already been configured (e.g. via a
    # command-line flag).
    config.default_formatter = 'doc'
  end

  # Print the 10 slowest examples and example groups at the end of the spec run, to help surface which specs are running
  # particularly slow.
  config.profile_examples = 10

  # Run specs in random order to surface order dependencies. If you find an order dependency and want to debug it, you
  # can fix the order by providing the seed (`--seed 1234`), which is printed after each run.
  config.order = :random

  # Seed global randomization in this process using the `--seed` CLI option. Setting this allows you to use `--seed` to
  # deterministically reproduce test failures related to randomization by passing the same `--seed` value as the one
  # that triggered the failure.
  Kernel.srand config.seed
end

spec/support/database_cleaner.rb
require 'database_cleaner'

RSpec.configure do |config|
  config.before(:suite) do
    DatabaseCleaner[:redis, { connection: :test }].strategy = :truncation

    DatabaseCleaner[:redis].clean_with :truncation
  end

  config.before(:each) do
    DatabaseCleaner[:redis].start
  end

  config.after(:each) do
    DatabaseCleaner[:redis].clean
  end
end

spec/ember_cli_deploy_rack_helper.rb
require 'rack/test'

root = File.expand_path '../../', __FILE__

# Configure the Rack environment.
ENV['RACK_ENV'] ||= 'test'

# Require the test object.
require 'ember/cli/deploy/rack'

# Define RSpec mixin.
module RSpecMixin
  include Rack::Test::Methods

  def app
    Ember::CLI::Deploy::Rack::Engine
  end

  def fixtures
    File.expand_path '../fixtures', __FILE__
  end
end

# Require the spec helper.
require 'spec_helper'

# Add additional requires below this line.

# Requires supporting Ruby files with custom matchers, macros etc. in `spec/support` and its subdirectories. Files
# matching `spec/**/*_spec.rb` are run as spec files by default. This means that files in `spec/support` that end in
# `_spec.rb` will both be required and run as specs, causing the specs to be run twice. It is recommended that you do
# not name files matching this glob to end with `_spec.rb`. You can configure this pattern with the `--pattern` option
# on the command line or in `~/.rspec`, `.rspec` or `.rspec-local`.

# The following line is provided for convenience purposes. It has the downside of increasing the boot-up time by
# auto-requiring all files in the `support` directory. Alternatively, in the individual `*_spec.rb` files, manually
# require only the support files necessary.
support = File.expand_path 'spec/support', root

Dir["#{support}/**/*.rb"].each { |file| require file }

RSpec.configure do |config|
  # Include the above defined RSpec mixin.
  config.include RSpecMixin
end

lib/ember/cli/deploy/rack/engine.rb
require 'logger'
require 'redis'
require 'sinatra/base'
require 'sinatra/config_file'
require 'sinatra/custom_logger'
require 'sinatra/reloader'

module Ember
  module CLI
    module Deploy
      module Rack # :nodoc:
        # The Engine class of `Ember::CLI::Deploy::Rack`, inherited from `Sinatra::Base`.
        class Engine < Sinatra::Base
          # === Constants ===

          TRAIL                 = File.expand_path('../../../../../../', __FILE__).freeze
          VIEWS                 = File.expand_path('views', TRAIL).freeze
          CONFIG                = 'config/settings.yml'.freeze
          KEY_PREFIX            = 'ember-cli-deploy-rack:index'.freeze
          ACTIVE_CONTENT_SUFFIX = 'current-content'.freeze
          REVISION_REGEXP       = '^[0-9a-f]{32}$'.freeze

          # === Settings ===

          set :root, TRAIL

          set :views, VIEWS

          set :key_prefix, KEY_PREFIX

          set :active_content_suffix, ACTIVE_CONTENT_SUFFIX

          set :revision, 'regexp' => REVISION_REGEXP

          set :redis, proc {
            defaults = Redis::Client::DEFAULTS.dup

            defaults.keys.each do |key|
              # Fill in defaults if needed
              defaults[key] = defaults[key].call if defaults[key].respond_to? :call
            end

            defaults
          }

          set :redis_client, proc {
            Redis.new settings.redis
          }

          set :debug, false

          set :version, Ember::CLI::Deploy::Rack::VERSION

          # === Extensions ===

          register Sinatra::ConfigFile

          config_file File.expand_path CONFIG

          # === Helpers ===

          helpers Sinatra::CustomLogger

          # === Configuration ===

          configure :development do
            register Sinatra::Reloader
          end

          # === Routes ===

          get '/debug' do
            if settings.debug
              @id       = 'debug'
              @title    = 'Debug'
              @settings = settings

              haml :debug
            else
              halt 403
            end
          end

          get '/*' do
            content_type 'text/html'

            logger.debug 'Processing request for `/*` route...'

            index params[:revision]
          end

          # === Private ===

          protected

          def index(revision)
            key = get_key revision

            logger.debug "Getting content from Redis with key: `#{key}`..."

            redis_client = settings.redis_client

            redis_client.get(key) || halt(404)
          end

          def get_key(revision)
            key = "#{settings.key_prefix}:#{settings.active_content_suffix}"

            if revision
              regexp = Regexp.new settings.revision['regexp'], 1

              if revision =~ regexp
                key = "#{settings.key_prefix}:#{revision}"
              else
                halt 400
              end
            end

            key
          end
        end
      end
    end
  end
end